### PR TITLE
OGG/OPUS support on iOS

### PR DESCRIPTION
**Pull request by Larpoux**
Hopefully this pull request is better than the previous.

I relay on another flutter plugin [flutter_ffmpeg plugin](https://pub.dev/packages/flutter_ffmpeg) instead of linking directly flutter_sound with ffmpeg.

This means that developers not interested by OGG/OPUS can continue to use flutter_sound without having to bother with ffmpeg.

### CODE REVIEW

**Comment by Larpoux** [comment]
There is a compatibility issue with my pull request:

Actually, on flutter_sound, the developer can specify a file name in "strartRecorder" and "stopRecorder".
something like : 
```dart
flutterSound.startRecorder(Platform.isIOS ? 'ios.aac' : 'android.aac');
```

This was really bad : 
- first: flutter_sound was supposed to create the file into the sdcard storage. This is not allowed anymore on Android 29.
- second: the developer could specify the file name, but not the complete path. He could not store his file where he wanted

I changed that : 
- If the developer does not specify any path, a temporary file is created in a temp directory
- If the developer wants to create his file somewhere else, he specify the full path.

Something like:
```dart
Directory tempDir = await getTemporaryDirectory ();
File fout = await File ('${tempDir.path}/flutter_sound-tmp.caf');
String path = await flutterSound.startRecorder(fout.path, codec: t_CODEC.CODEC_AAC,);
```
I modified the README accordingly.

I really do not want to keep the previous behaviour which was a big mistake, I think.

File: README.md [hunk 1]
```diff
@@ -107,21 +103,27 @@ result.then(path) {
...
 - CAF/OPUS
+-  OGG/OPUS
+
+Recently, Apple added a support for encoding with the standard OPUS codec. Unfortunatly, Apple encapsulates its data in its own proprietary envelope : CAF. This is really stupid, this is Apple. If you need to record with regular OGG/OPUS you must add `flutter_ffmpeg` to your dependencies.
+Please, look to the [flutter_ffmpeg plugin README](https://pub.dev/packages/flutter_ffmpeg) for instructions for how to include this plugin into your app
 
 To encode with OPUS you do the following :
 ```dart
-await flutterSound.startRecorder(codec: t_CODEC.CODEC_CAF_OPUS,)
+await flutterSound.startRecorder(foot.path, codec: t_CODEC.CODEC_OPUS,)
```
Comment (hyochan) [inline README.md]: what is this `foot.path` stands for?

File: README.md [hunk 2]
```diff
@@ -156,7 +158,9 @@ You must wait for the return value to complete before attempting to add any list
 to ensure that the player has fully initialised.
 
 ```dart
-Future<String> result = await flutterSound.startPlayer(null);
+Directory tempDir = await getTemporaryDirectory ();
```
Comment (hyochan) [inline README.md]: Please remove the`space` before calling the function `()`.

File: android/build.gradle [hunk 3]
```diff
@@ -1,6 +1,8 @@
 group 'com.dooboolab.fluttersound'
 version '1.0-SNAPSHOT'
 
+ext.flutterFFmpegPackage = 'audio'
```
Comment (hyochan) [inline android/build.gradle]: Could you kindly explain why this is necessary?

File: example/android/app/build.gradle [hunk 4]
```diff
@@ -34,7 +34,7 @@ android {
     defaultConfig {
         // TODO: Specify your own unique Application ID (https://developer.android.com/studio/build/application-id.html).
         applicationId "com.dooboolab.fluttersoundexample"
-        minSdkVersion 16
+        minSdkVersion 24
```
Comment (hyochan) [inline example/android/app/build.gradle]: Why is this necessary to change the minSdkVersion in `example`?

File: example/lib/main.dart [hunk 5]
```diff
@@ -42,6 +42,8 @@ class _MyAppState extends State<MyApp> {
   double maxDuration = 1.0;
   t_MEDIA _media = t_MEDIA.FILE;
   t_CODEC _codec = t_CODEC.CODEC_AAC;
+  bool _encoderSupported = true; // Optimist
```
Comment (hyochan) [inline example/lib/main.dart]: Is there any better way to check this status dynamically?

File: ios/Classes/FlutterSoundPlugin.m [hunk 6]
```diff
@@ -282,7 +282,7 @@ - (void)startRecorder
   if ([path class] == [NSNull class]) {
     audioFileURL = [NSURL fileURLWithPath:[GetDirectoryOfType_FlutterSound(NSCachesDirectory) stringByAppendingString:defaultExtensions[codec] ]];
   } else {
-    audioFileURL = [NSURL fileURLWithPath: [GetDirectoryOfType_FlutterSound(NSCachesDirectory) stringByAppendingString:path]];
```
Comment (hyochan) [inline ios/Classes/FlutterSoundPlugin.m]: So you don't like the cache directory? How about supporting options to choose cache or file?

File: lib/flutter_sound.dart [hunk 7]
```diff
@@ -192,7 +281,11 @@ class FlutterSound {
 
       String result = await _channel.invokeMethod('startRecorder', param);
       _setRecorderCallback();
-        _audio_state = t_AUDIO_STATE.IS_RECORDING;
+      _audio_state = t_AUDIO_STATE.IS_RECORDING;
```
Comment (hyochan) [inline lib/flutter_sound.dart]: _audio_state => _audioState

File: lib/flutter_sound.dart [hunk 8]
```diff
@@ -192,7 +281,11 @@ class FlutterSound {
...
+      // if the caller wants OGG/OPUS we must remux the temporary file
+      if ( (result != null) && ogg_opus) {
+         return savUri;
```
Comment (hyochan) [inline lib/flutter_sound.dart]: savUri => saveUri

File: lib/flutter_sound.dart [hunk 9]
```diff
@@ -43,9 +45,14 @@ final List<String> paths =
 
 class FlutterSound {
   static const MethodChannel _channel = const MethodChannel('flutter_sound');
+  static const MethodChannel _FFmpegChannel = const MethodChannel('flutter_ffmpeg');
   static StreamController<RecordStatus> _recorderController;
   static StreamController<double> _dbPeakController;
   static StreamController<PlayStatus> _playerController;
+  static bool ogg_opus = false; // Set by startRecorder when the user wants to record an ogg/opus
```
Comment (hyochan) [inline lib/flutter_sound.dart]: ogg_opus => isOppOpus ?

File: lib/flutter_sound.dart [hunk 10]
```diff
@@ -43,9 +45,14 @@ final List<String> paths =
...
+  static String savUri; // Used by startRecorder/stopRecorder to keep the caller wanted uri
```
Comment (hyochan) [inline lib/flutter_sound.dart]: savUri => savedUri

File: lib/flutter_sound.dart [hunk 11]
```diff
@@ -209,11 +302,33 @@ class FlutterSound {
     _audio_state = t_AUDIO_STATE.IS_STOPPED;
     _removeRecorderCallback();
     _removeDbPeakCallback();
+
+    if (  ogg_opus) {
```
Comment (hyochan) [inline lib/flutter_sound.dart]: Please remove space before `ogg_opus`.

File: ios/Classes/FlutterSoundPlugin.m (hunk 6 above)
Comment (Larpoux) [reply ios/Classes/FlutterSoundPlugin.m]: My idea is that the recorded file will be stored in the cache directory if the user does not specify a file path.
And it will be stored in the provided path if the user specify a path.

I really think that it is not necessary for the user to specify just a temporary file name and not the full path. Of course we can still do that for backward compatibility. Just tell me.

File: example/android/app/build.gradle (hunk 4 above)
Comment (Larpoux) [reply example/android/app/build.gradle]: This is because flutter_ffmpeg works on minSdkVersion 24 (or later).
If the user does not need OPUS he can continue to use SDK 16.
But if it is necessary to have SDK 24 for ffmpeg.

The example needs OPUS to test how it works.

Note: Actually, ffmpeg is not used on Android. But we probably will use it later, for example to use `lame` for encoding MP3

File: android/build.gradle (hunk 3 above)
Comment (Larpoux) [reply android/build.gradle]: Actually this is not really necessary : Android still does not use ffmpeg.
(It will probably be used later, for recording MP3, for example).

I am commenting out this line in the current version.

File: ios/Classes/FlutterSoundPlugin.m (hunk 6 above)
Comment (hyochan) [reply ios/Classes/FlutterSoundPlugin.m]: Oh! I think your idea is absolutely great~

File: example/android/app/build.gradle (hunk 4 above)
Comment (hyochan) [reply example/android/app/build.gradle]: Could you also provide this detail in the readme for other users of flutter_sound?

**Comment by hyochan** [comment]
@Larpoux How is this going my friend?

**Comment by Larpoux** [comment]
I am **very** happy : my tests on various SDK version, both on Android and iOS, both with flutter embedded V1 and flutter embedded V2 are very positive. I think it will be possible for me to do a pull request this evening.

Now, I request recording authorisation on Dart side instead of Java. It works great, and suppress many many incompatibilities issues that we had before.

My problem is that I work, now, on a new branch issued from my last pull request, #199 ffmpeg.
If you reject my pull request  "OGG/OPUS support on iOS #199" I will back-port my current branch to the flutter_sound master branch.

I remember that you asked me to work on #197. I will try to look to it tomorrow, but right now I prefer to finish #203 .

For the two bugs report I did yesterday, I will take care of them, of course. It was just for not loosing those little bugs in case you reject my other pull requests.

**Comment by hyochan** [comment]
@Larpoux No rush my friend. I just wonder if this work is complete. Thanks for your time.

File: example/android/app/build.gradle (hunk 4 above)
Comment (Larpoux) [reply example/android/app/build.gradle]: flutter_ffmpeg LTS Release - minAndroidSdk 16 (instead of 24)

I was able to build the Example App with MinAndroidSdk 16, using the LTS Release of flutter_ffmpeg instead of the Main Release

**Comment by Larpoux** [comment]
Ok Hyo, I really thin that this pull request can be integrated in your Master Branch.

I use it successfully in my own App. Everything run correctly.

**Comment by MATTYGILO** [comment]
Is this a branch or in the new code
